src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -628,110 +629,72 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
-                    {
-                        elementNames = ArrayBuilder<string>.GetInstance(arguments.Count);
-                        for (int j = 0; j < i; j++)
-                        {
-                            elementNames.Add(TupleTypeSymbol.UnderlyingMemberName(j));
-                        }
-                        elementNames.Add(name);
-                    }
-                }
-
-                var boundArgument = BindValue(argumentSyntax.Expression, diagnostics, BindValueKind.RValue);
+                BoundExpression boundArgument = BindValue(argumentSyntax.Expression, bag, BindValueKind.RValue);
                 boundArguments.Add(boundArgument);
 
                 // PROTOTYPE: need to report distinc errors for tuples
-                var elementType = GetAnonymousTypeFieldType(boundArgument, argumentSyntax, diagnostics, ref hasErrors);
+                bool hasErrors = false;
+                TypeSymbol elementType = GetAnonymousTypeFieldType(boundArgument, argumentSyntax, bag, ref hasErrors);
                 elementTypes.Add(elementType);
             }
+            uniqueFieldNames.Free();
 
             if (countOfExplicitNames != 0 && countOfExplicitNames != elementTypes.Count)
             {
-                hasErrors = true;
-                Error(diagnostics, ErrorCode.ERR_TupleExplicitNamesOnAllMembersOrNone, node);
-            }
-            var elements = elementTypes.ToImmutableAndFree();
-            if (elements.Length < 2 || elements.Length > 7)
-            {
-                // PROTOTYPE
-                diagnostics.Add(ErrorCode.ERR_PrototypeNotYetImplemented, node.Location);
-                return BadExpression(node);
+                Error(bag, ErrorCode.ERR_TupleExplicitNamesOnAllMembersOrNone, node);
             }
 
+            ImmutableArray<TypeSymbol> elements = elementTypes.ToImmutableAndFree();
+
             var type = new TupleTypeSymbol(
                 elements,
                 elementNames == null ?
-                                default(ImmutableArray<string>) :
-                                elementNames.ToImmutableAndFree(),
+                    default(ImmutableArray<string>) :
+                    elementNames.ToImmutableAndFree(),
                 node,
                 this,
-                diagnostics);
-
-            var ctor = (MethodSymbol)GetWellKnownTypeMember(this.Compilation, TupleTypeSymbol.GetTupleCtor(elements.Length), diagnostics, syntax: node);
-            if ((object)ctor != null)
-            {
-                ctor = ctor.AsMember(type.UnderlyingTupleType);
-            }
-            else
-            {
-                hasErrors = true;
-            }
+                bag);
 
-            return new BoundTupleCreationExpression(node, ctor, boundArguments.ToImmutableAndFree(), type, hasErrors);
+            diagnostics.AddRange(bag);
+            return new BoundTupleCreationExpression(node, boundArguments.ToImmutableAndFree(), type, bag.HasAnyErrors());
```

Comment:
Calling HasAnyErrors during initial binding is a bad idea, it resolves lazy diagnostics, which, if we do this too early, can cause cycles. Using it as a way of calculating HasErrors is also wrong because we should never mark nodes as bad because obsolete errors. Please go back to the original way of tracking HasErrors. We can talk more offline about when we should mark nodes with that flag.